### PR TITLE
drop python<=3.7 support

### PR DESCRIPTION
**Pull request by kloczek**
According to https://endoflife.date/python python 3.7 has been EOSed 27 Jun 2023.
Filter all code over `pyupgracde --py38-plus`.

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -23,7 +23,7 @@
     'pytz',
     # install requirements depending on python version
     # see https://www.python.org/dev/peps/pep-0508/#environment-markers
-    'backports.zoneinfo; python_version == "3.7" or python_version == "3.8"',
+    'backports.zoneinfo; python_version <= "3.8"',
```
Comment (stevepiercy) [inline setup.py]: This change would allow Python 3.6 and earlier. Will that actually work? If it cannot be confirmed, then I would revert this change.

**Review by stevepiercy — CHANGES_REQUESTED** [review]
I don't think this is correct. icalendar v6 supports both timezone implementations `pytz` and `zoneinfo` and it will continue to support Python 3.7. This will allow an upgrade path for those stuck on old Pythons.

- https://github.com/collective/icalendar/issues/630
- https://github.com/collective/icalendar/issues/513

Some of the other changes in this PR could be useful, if Python 3.7 supports it.

File: setup.py (hunk 1 above)
Comment (kloczek) [reply setup.py]: Nope .. all because `python_requires=">=3.8"`

Comment (kloczek) [reply setup.py]: Actually that part is buggy `backports.zoneinfo` dependency should be `'backports.zoneinfo; python_version <= "3.9"',`

File: .github/workflows/tests.yml [hunk 2]
```diff
@@ -17,13 +17,13 @@ jobs:
...
         config:
         # [Python version, tox env]
-        - ["3.7",   "py37"]
         - ["3.8",   "py38"]
         - ["3.9",   "py39"]
         - ["3.10",  "py310"]
         - ["pypy-3.9", "pypy3"]
         - ["3.10",  "docs"]
         - ["3.11",  "py311"]
+        - ["3.12",  "py312"]
```
Comment (niccokunzmann) [inline .github/workflows/tests.yml:26]: nice!

File: src/icalendar/tests/test_with_doctest.py [hunk 3]
```diff
@@ -73,6 +73,3 @@ def test_documentation_file(document, zoneinfo_only, env_for_doctest):
 
 def test_can_import_zoneinfo(env_for_doctest):
     """Allow importing zoneinfo for tests."""
-    import pytz
```
Comment (niccokunzmann) [inline src/icalendar/tests/test_with_doctest.py:76]: These three need to stay. The others can go. They are part of the test.

File: setup.py (hunk 1 above)
Comment (niccokunzmann) [reply setup.py]: OUTDATED...

```
Python 3.9.18 (main, Oct 23 2023, 13:20:13) 
>>> import zoneinfo
```

I would say `backports.zoneinfo; python_version < "3.9"`

I wonder if that would fox this in the doctest.
```
Expected:
    ZoneInfo(key='Europe/Vienna')
Got:
    zoneinfo.ZoneInfo(key='Europe/Vienna')
```

**Review by niccokunzmann — CHANGES_REQUESTED** [review]
Hi! Thanks for using this tool to update the codebase! It is really useful!
I added a few points that would be hard to find.
I wonder why the tests break in that place. It is unusual to have this inconsistent. I added my idea of why.

What do you think? How would you like to proceed? Thanks again for taking this up!

File: src/icalendar/tests/test_with_doctest.py (hunk 3 above)
Comment (niccokunzmann) [reply src/icalendar/tests/test_with_doctest.py:76]: Hm. One could probably  use them in there....

**Comment by niccokunzmann** [comment]
Thanks! If you could...
- add yourself to the contributors' list
- add a changelog entry under "breaking change"

**Comment by niccokunzmann** [comment]
I wonder why py39 fails... I will have a look.

File: setup.py (hunk 1 above)
Comment (niccokunzmann) [reply setup.py]: https://github.com/kloczek/icalendar/pull/1 resolves this.

**Comment by kloczek** [comment]
I've corrected last commit from "<= 3.9" to "< 3.9".

**Comment by niccokunzmann** [comment]
@kloczek Thanks for creating this PR! I am planning on releasing v6 and including this PR in it. So, I hope we can get this merged. :rocket: 
Where would you like to go with this?

**Comment by niccokunzmann** [comment]
The only changes I would request are these:

- [ ] Do not delete the test case https://github.com/collective/icalendar/pull/634#discussion_r1649233685
- [ ] add a changelog entry

**Review by stevepiercy — CHANGES_REQUESTED** [review]
Wait for #647 to be merged. Then update `CHANGES.rst`. See https://github.com/collective/icalendar/pull/647#pullrequestreview-2134619227

**Comment by kloczek** [comment]
As I;ve mentioned as globally it is declared to exit with error on any try to build on python < 3.8 it is not possible pass that rule on generate .whl archive where are placed install time dependencies.
From my package icalendar-5.0.13.dist-info/METADATA:
```console
Metadata-Version: 2.1
Name: icalendar
Version: 5.0.13
Summary: iCalendar parser/generator
Home-page: https://github.com/collective/icalendar
Author: Plone Foundation
Author-email: plone-developers@lists.sourceforge.net
License: BSD-2-Clause
Keywords: calendar calendaring ical icalendar event todo journal recurring
Classifier: Development Status :: 5 - Production/Stable
Classifier: Intended Audience :: Developers
Classifier: License :: OSI Approved :: BSD License
Classifier: Operating System :: OS Independent
Classifier: Programming Language :: Python :: 3.8
Classifier: Programming Language :: Python :: 3.9
Classifier: Programming Language :: Python :: 3.10
Classifier: Programming Language :: Python :: 3.11
Classifier: Programming Language :: Python :: 3.12
Classifier: Programming Language :: Python :: Implementation :: CPython
Classifier: Programming Language :: Python :: Implementation :: PyPy
Requires-Python: >=3.8                      <<<<<<<<==== HERE.
License-File: LICENSE.rst
Requires-Dist: python-dateutil
Requires-Dist: pytz                              <<<<<<<===== AAND HERE
Requires-Dist: backports.zoneinfo ; python_version < "3.9" <<== AND HERE
Provides-Extra: test
```

**Comment by niccokunzmann** [comment]
Cool, yes. That is alright.

If you address these two concerns, I would merge it:

- [ ] Do not delete the test case https://github.com/collective/icalendar/pull/634#discussion_r1649233685
- [ ] add a changelog entry

File: src/icalendar/tests/test_with_doctest.py (hunk 3 above)
Comment (niccokunzmann) [reply src/icalendar/tests/test_with_doctest.py:76]: #650 obsoletes this change, will create a merge conflict. We can thus just merge it and it will not get lost.

**Comment by niccokunzmann** [comment]
Thanks @kloczek for your contribution!
For others: The changelog entry is not mandatory but a good indicator.

**Comment by kloczek** [comment]
Thank you 👍 
Sorry but I have some shortages of time to finish your requests.
